governance/program/src/state/proposal_transaction.rs

```diff
@@ -95,62 +97,74 @@ pub struct ProposalInstructionV2 {
     /// The option index the instruction belongs to
     pub option_index: u16,
 
-    /// Unique instruction index within it's parent Proposal
-    pub instruction_index: u16,
+    /// Unique transaction index within it's parent Proposal
+    pub transaction_index: u16,
 
     /// Minimum waiting time in seconds for the  instruction to be executed once proposal is voted on
     pub hold_up_time: u32,
 
-    /// Instruction to execute
-    /// The instruction will be signed by Governance PDA the Proposal belongs to
+    /// Instructions to execute
+    /// The instructions will be signed by Governance PDA the Proposal belongs to
     // For example for ProgramGovernance the instruction to upgrade program will be signed by ProgramGovernance PDA
-    pub instruction: InstructionData,
+    // All instructions will be executed withing a single transaction
+    pub instructions: Vec<InstructionData>,
 
     /// Executed at flag
     pub executed_at: Option<UnixTimestamp>,
 
     /// Instruction execution status
-    pub execution_status: InstructionExecutionStatus,
+    pub execution_status: TransactionExecutionStatus,
 }
 
-impl AccountMaxSize for ProposalInstructionV2 {
+impl AccountMaxSize for ProposalTransactionV2 {
     fn get_max_size(&self) -> Option<usize> {
-        Some(self.instruction.accounts.len() * 34 + self.instruction.data.len() + 91)
+        let instructions_size = self
+            .instructions
+            .iter()
+            .map(|i| i.accounts.len() * 34 + i.data.len())
+            .sum::<usize>()
+            + 4;
+
+        Some(instructions_size + 91)
     }
 }
 
-impl IsInitialized for ProposalInstructionV2 {
+impl IsInitialized for ProposalTransactionV2 {
     fn is_initialized(&self) -> bool {
-        self.account_type == GovernanceAccountType::ProposalInstructionV2
+        self.account_type == GovernanceAccountType::ProposalTransactionV2
     }
 }
 
-impl ProposalInstructionV2 {
+impl ProposalTransactionV2 {
     /// Serializes account into the target buffer
     pub fn serialize<W: Write>(self, writer: &mut W) -> Result<(), ProgramError> {
-        if self.account_type == GovernanceAccountType::ProposalInstructionV2 {
+        if self.account_type == GovernanceAccountType::ProposalTransactionV2 {
             BorshSerialize::serialize(&self, writer)?
         } else if self.account_type == GovernanceAccountType::ProposalInstructionV1 {
+            if self.instructions.len() != 1 {
+                panic!("Multiple instructions are not supported by ProposalInstructionV1")
```

Comment:
This already returns a `Result<(), ProgramError>` so you're better off returning an error here.  Then again, you should probably never arrive at this point, right?

Reply:
Correct, we don't mutate `ProposalTransaction`. If we want to change it we replace the entire account, so this check is just in case something would change in future versions